device/iot-device-samples/device-reconnection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceClientManager.java

```diff
@@ -0,0 +1,146 @@
+package samples.com.microsoft.azure.sdk.iot;
+
+import com.microsoft.azure.sdk.iot.device.DeviceClient;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeCallback;
+import com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason;
+import com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus;
+import lombok.experimental.Delegate;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.COMMUNICATION_ERROR;
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+
+@Slf4j
+public class DeviceClientManager implements IotHubConnectionStatusChangeCallback {
+    private enum ConnectionStatus {
+        DISCONNECTED, CONNECTING, CONNECTED
+    }
+
+    private static final Object lock = new Object();
+    private static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 10;
+    private final boolean autoReconnectOnDisconnected;
+    private ConnectionStatus connectionStatus;
+    private IotHubConnectionStatusChangeCallback suppliedConnectionStatusChangeCallback;
+
+    private interface DeviceClientNonDelegatedFunction {
+        void open();
+        void closeNow();
+        void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext);
+    }
+
+    // The methods defined in the interface DeviceClientNonDelegatedFunction will be called on DeviceClientManager, and not on DeviceClient.
+    @Delegate(excludes = DeviceClientNonDelegatedFunction.class)
+    private final DeviceClient client;
+
+    DeviceClientManager(DeviceClient deviceClient, boolean autoReconnectOnDisconnected) {
+        this.connectionStatus = ConnectionStatus.DISCONNECTED;
+        this.client = deviceClient;
+        this.autoReconnectOnDisconnected = autoReconnectOnDisconnected;
+    }
+
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext) {
+        this.suppliedConnectionStatusChangeCallback = callback;
+        this.client.registerConnectionStatusChangeCallback(this, callbackContext);
```

Comment:
`this.client.registerConnectionStatusChangeCallback(this, callbackContext);` should be called only once in constructor.

Reply:
Reconnection depends on the connection status returned by the connection status change callback; so behavior-wise, we set it only if the application sets the connection status change callback